scripts/ingest.py

```diff
@@ -162,44 +164,52 @@ async def on_shutdown(server: IngestDeviceServer) -> None:
-        sys.exit(1)
-    cbf_channels = cbf_attr['n_chans']
-    if args.output_channels is None:
-        args.output_channels = Range(0, cbf_channels)
-    if args.sd_output_channels is None:
-        args.sd_output_channels = Range(0, cbf_channels)
-    # If no continuum product is selected, set continuum factor to 1 since
-    # that effectively disables the alignment checks.
-    continuum_factor = args.continuum_factor if args.l0_continuum_spead else 1
-    # TODO: determine an appropriate value for guard
-    channel_ranges = ChannelRanges(
-        args.servers, args.server_id - 1,
-        cbf_channels, continuum_factor, args.sd_continuum_factor,
-        len(args.cbf_spead), args.guard_channels, args.output_channels, args.sd_output_channels)
-    context = accel.create_some_context(interactive=False)
-    server = IngestDeviceServer(args, channel_ranges, cbf_attr, context, args.host, args.port)
+    telstate = await get_async_telstate(args.telstate_endpoint)
+    telstate_cbf = await cbf_telstate_view(telstate, args.cbf_name)
+    async with katsdpmodels.fetch.aiohttp.TelescopeStateFetcher(telstate) as fetcher:
+        system_attrs = await SystemAttrs.create(fetcher, telstate_cbf, args.antenna_mask)
+        cbf_channels = system_attrs.cbf_attr['n_chans']
+        if args.output_channels is None:
+            args.output_channels = Range(0, cbf_channels)
+        if args.sd_output_channels is None:
+            args.sd_output_channels = Range(0, cbf_channels)
+        # If no continuum product is selected, set continuum factor to 1 since
+        # that effectively disables the alignment checks.
+        continuum_factor = args.continuum_factor if args.l0_continuum_spead else 1
+        # TODO: determine an appropriate value for guard
+        channel_ranges = ChannelRanges(
+            args.servers, args.server_id - 1,
+            cbf_channels, continuum_factor, args.sd_continuum_factor,
+            len(args.cbf_spead), args.guard_channels, args.output_channels, args.sd_output_channels)
+        context = accel.create_some_context(interactive=False)
+        server = IngestDeviceServer(args, telstate_cbf, channel_ranges, system_attrs, context,
+                                    args.host, args.port)
 
     loop.add_signal_handler(signal.SIGINT, lambda: loop.create_task(on_shutdown(server)))
     loop.add_signal_handler(signal.SIGTERM, lambda: loop.create_task(on_shutdown(server)))
-    loop.run_until_complete(server.start())
+    await server.start()
     logger.info("Started katsdpingest server.")
     with katsdpservices.start_aiomonitor(loop, args, locals()):
-        loop.run_until_complete(server.join())
+        await server.join()
+        telstate.backend.close()
+        await telstate.backend.wait_closed()
     logger.info("Shutdown complete")
-    loop.close()
 
 
 if __name__ == '__main__':
-    main()
+    loop = asyncio.get_event_loop()
```

Comment:
Out of interest, is this the latest way to call async scripts?

Reply:
Not exactly. Python 3.7 introduced `asyncio.run(main())` to deal with this boilerplate stuff. But we're still using Python 3.6.

Reply:
Thanks for refreshing my memory!